File: agent11-frontend /src/api/knowledge.ts
import apiClient from './client'

export interface KnowledgeFile {
  id?: string
  filename: string
  path: string
  content?: string
  isFolder?: boolean
  createdAt: string
  updatedAt: string
  children?: KnowledgeFile[]
}

export interface GetKnowledgeResponse {
  success: boolean
  data?: KnowledgeFile[]
  error?: string
}

export interface GetFileResponse {
  success: boolean
  data?: KnowledgeFile
  error?: string
}

export interface SearchKnowledgeRequest {
  query: string
  limit?: number
}

export interface SearchKnowledgeResponse {
  success: boolean
  data?: {
    filename: string
    content: string
    score: number
  }[]
  error?: string
}

export interface FolderRequest {
  name: string
  parentPath?: string
}

export const knowledgeApi = {
  getFiles: async (path?: string): Promise<GetKnowledgeResponse> => {
    const url = path ? `/knowledge?path=${encodeURIComponent(path)}` : '/knowledge'
    const response = await apiClient.get<GetKnowledgeResponse>(url)
    return response.data
  },

  getFile: async (filename: string): Promise<GetFileResponse> => {
    const response = await apiClient.get<GetFileResponse>(`/knowledge/${encodeURIComponent(filename)}`)
    return response.data
  },

  uploadFile: async (file: File, folderPath?: string): Promise<{ success: boolean; data?: { filename: string }; error?: string }> => {
    const formData = new FormData()
    formData.append('file', file)
    if (folderPath) {
      formData.append('path', folderPath)
    }
    // Use apiClient with content-type unset so axios auto-generates boundary
    const response = await apiClient.post('/knowledge/upload', formData, {
      headers: {
        // Override Content-Type to let axios set it with proper boundary
        'Content-Type': undefined,
      },
    })
    return response.data
  },

  deleteFile: async (filename: string): Promise<{ success: boolean; error?: string }> => {
    const response = await apiClient.delete(`/knowledge/${encodeURIComponent(filename)}`)
    return response.data
  },

  updateFile: async (filename: string, content: string): Promise<{ success: boolean; error?: string }> => {
    const response = await apiClient.put(`/knowledge/${encodeURIComponent(filename)}`, { content })
    return response.data
  },

  search: async (data: SearchKnowledgeRequest): Promise<SearchKnowledgeResponse> => {
    const response = await apiClient.post<SearchKnowledgeResponse>('/knowledge/search', data)
    return response.data
  },

  createFolder: async (data: FolderRequest): Promise<{ success: boolean; error?: string }> => {
    const response = await apiClient.post('/knowledge/folder', data)
    return response.data
  },

  deleteFolder: async (path: string): Promise<{ success: boolean; error?: string }> => {
    const response = await apiClient.delete(`/knowledge/folder/${encodeURIComponent(path)}`)
    return response.data
  },
}

export default knowledgeApi

File: agent11-backend/app/api/routes/knowledge.py
"""知识库 API"""
from fastapi import APIRouter, UploadFile, File, Query, Form
from pydantic import BaseModel
import structlog

from app.knowledge.manager import KnowledgeManager

logger = structlog.get_logger()

router = APIRouter()


class SearchRequest(BaseModel):
    query: str
    limit: int = 5


class FolderRequest(BaseModel):
    name: str
    parentPath: str = ""


km = KnowledgeManager()


@router.get("/")
async def list_knowledge(path: str = Query("")):
    """列出知识库文件"""
    items = await km.list_files(path)
    return {"success": True, "data": items}


@router.get("/{filename:path}")
async def get_file(filename: str):
    """获取文件内容"""
    content = await km.get_file_content(filename)
    if content is None:
        return {"success": False, "error": "File not found"}
    return {
        "success": True,
        "data": {
            "filename": filename.split("/")[-1],
            "path": filename,
            "content": content,
            "isFolder": False,
            "createdAt": None,
            "updatedAt": None,
        }
    }


@router.post("/upload")
async def upload_file(
    file: UploadFile = File(...),
    path: str = Form(""),
):
    """上传文件"""
    content = await file.read()
    success, result = await km.upload_file(file.filename, content, path)

    if success:
        return {"success": True, "data": {"filename": result}}
    else:
        return {"success": False, "error": result}


class UpdateFileRequest(BaseModel):
    content: str


@router.put("/{filename:path}")
async def update_file(filename: str, body: UpdateFileRequest):
    """更新文件内容"""
    success = await km.save_file(filename, body.content)
    if success:
        return {"success": True}
    else:
        return {"success": False, "error": "Failed to save file"}


@router.delete("/{filename:path}")
async def delete_file(filename: str):
    """删除文件"""
    success = await km.delete_file(filename)
    if success:
        return {"success": True}
    else:
        return {"success": False, "error": "Failed to delete file"}


@router.post("/search")
async def search_knowledge(request: SearchRequest):
    """语义搜索知识库"""
    results = await km.search(request.query, request.limit)
    return {
        "success": True,
        "data": [
            {
                "filename": r["filename"],
                "content": r["content"],
                "score": r["score"]
            }
            for r in results
        ]
    }


@router.post("/folder")
async def create_folder(request: FolderRequest):
    """创建文件夹"""
    success = await km.create_folder(request.name, request.parentPath)
    if success:
        return {"success": True}
    else:
        return {"success": False, "error": "Failed to create folder"}


@router.delete("/folder/{path:path}")
async def delete_folder(path: str):
    """删除文件夹"""
    success = await km.delete_folder(path)
    if success:
        return {"success": True}
    else:
        return {"success": False, "error": "Failed to delete folder"}